### PR TITLE
move some unsafe code inside an unsafe{} boundary

### LINKED ISSUE #481: Internal "unsoundness" in `sign_ecdsa_with_noncedata_pointer`
While `sign_ecdsa_with_noncedata_pointer` is private so consumers can't cause UB in safe code it's still considered wrong to not mark the function `unsafe` because it's confusing. From my understanding such issue caused horrible flamewar at `actix-web` crate.

Given that the function is called either with `&[u8; 32]` or null pointer I propose to change it to accept `Option<&[u8; 32]>` instead. Internally, it does the same thing thanks to layout optimizations but is safer. We supply null pointer if there is `None` and the pointer obtained from reference if there is `Some`.

### PR DESCRIPTION
**Pull request by apoelstra**
An internal function had a non-unsafe signature but could be called
with data that would cause it to exhibit UB. Move the unsafety inside
of the function so that the function signature now enforces soundness.

Fixes #481

### CODE REVIEW

**Comment by apoelstra** [comment]
Not sure if this is as tight as you were hoping for @Kixunil

**Review by tcharding — APPROVED** [review]
FWIW looks ok to me.

ACK 1242e7131a0ca8f6eb7943a6e6d5d7acfe12b2ba

**Comment by apoelstra** [comment]
Interesting. I saw that clippy failure but ignored it since it had nothing to do with my code ... I guess it's a change in clippy and we need to fix it for CI to pass.

**Comment by tcharding** [comment]
I don't think it was a change in clippy because we use `clippy.toml` to set the version. I think this was caused by the same "github ci doesn't re-run when master changes" thing we've been hitting a bunch lately. *cough* bors :)

**Review by sanket1729 — DISMISSED** [review]
ACK 29ce1cc0e4dd04688d3aadfa289bfae51d7832f0

File: src/ecdsa/mod.rs [hunk 1]
```diff
@@ -257,10 +257,14 @@ impl<C: Signing> Secp256k1<C> {
...
         sk: &SecretKey,
-        noncedata_ptr: *const ffi::types::c_void,
+        noncedata: Option<&[u8; 32]>,
     ) -> Signature {
         unsafe {
             let mut ret = ffi::Signature::new();
+            let noncedata_ptr = match noncedata {
+                Some(arr) => arr.as_c_ptr() as *const _,
+                None => ptr::null(),
+            };
```
Comment (Kixunil) [inline src/ecdsa/mod.rs:267]: Oh, you didn't change the signature of `secp256k1_ecdsa_sign` to avoid hassle with `-sys` breaking release?

**Review by Kixunil — APPROVED** [review]
ACK 29ce1cc0e4dd04688d3aadfa289bfae51d7832f0

But maybe you want commits in the opposite order?

**Comment by tcharding** [comment]
Now I'm confused, I did a clippy PR https://github.com/rust-bitcoin/rust-secp256k1/pull/484 that includes additional fixes to this PR but this is green?

**Comment by apoelstra** [comment]
CI only runs on the top commit -- and the top commit of this PR has the clippy fixes.

**Comment by apoelstra** [comment]
Merged #484 and rebased. Now there are no clippy fixes in this PR.

**Review by elichai — APPROVED** [review]
LGTM

**Comment by apoelstra** [comment]
Merged -- the one remaining commit was identical to the one ACKed by Sanket and Kixunil, and while elichai did not specify a commit, there is really only one that was ever in this PR.

**Comment by tcharding** [comment]
> CI only runs on the top commit 

Oh I'm a goose.

**Pull request merged**